Source/WebCore/dom/Element.cpp

```diff
@@ -4740,6 +4742,52 @@ bool Element::isSpellCheckingEnabled() const
     return true;
 }
 
+bool Element::isWritingSuggestionsEnabled() const
+{
+    // If none of the following conditions are true, then return `false`.
+
+    // `element` is an `input` element whose `type` attribute is in either the
+    // `Text`, `Search`, `URL`, `Email` state and is `mutable`.
+    auto isEligibleInputElement = [&] {
+        RefPtr input = dynamicDowncast<HTMLInputElement>(this);
+        if (!input)
+            return false;
+
+        auto isAllowedField = input->isTextField() || input->isSearchField() || input->isURLField() || input->isEmailField();
+        if (!isAllowedField)
+            return false;
+
+        return true;
+    }();
+
+    // `element` is a `textarea` element that is `mutable`.
+    auto isTextArea = is<HTMLTextAreaElement>(*this);
+
+    // `element` is an `editing host` or is `editable`.
+    auto isEditable = hasEditableStyle() || isContentEditable();
+
+    if (!isEligibleInputElement && !isTextArea && !isEditable)
+        return false;
+
+    // If `element` has an 'inclusive ancestor' with a `writingsuggestions` content attribute that's
+    // not in the `default` state and the nearest such ancestor's `writingsuggestions` content attribute
+    // is in the `false` state, then return `false`.
+
+    for (auto* ancestor = this; ancestor; ancestor = ancestor->parentOrShadowHostElement()) {
```

Comment:
I'm pretty sure the specification was only gonna walk the normal tree. Did that change?